File: sdk/python/jamjet/cloud/redaction.py
"""PII redaction for JamJet Cloud SDK.

Uses Presidio (presidio-analyzer + spacy en_core_web_lg) when installed.
Falls back to compiled regex patterns otherwise.
"""
from __future__ import annotations

import re
from typing import Any

try:
    from presidio_analyzer import AnalyzerEngine
    from presidio_anonymizer import AnonymizerEngine
    from presidio_anonymizer.entities import OperatorConfig

    _presidio_available = True
    _analyzer: Any = None
    _anonymizer: Any = None
except ImportError:
    _presidio_available = False
    _analyzer = None
    _anonymizer = None


_REGEX_PATTERNS: dict[str, re.Pattern[str]] = {
    "EMAIL_ADDRESS": re.compile(
        r"[a-zA-Z0-9._%+\-]+@[a-zA-Z0-9.\-]+\.[a-zA-Z]{2,}"
    ),
    "CREDIT_CARD": re.compile(r"\b(?:\d[\s\-]?){13,15}\d\b"),
    "US_SSN": re.compile(r"\b\d{3}[-\s]?\d{2}[-\s]?\d{4}\b"),
    "PHONE_NUMBER": re.compile(
        r"\b(?:\+?1[-.\s]?)?\(?\d{3}\)?[-.\s]?\d{3}[-.\s]?\d{4}\b"
    ),
    "IP_ADDRESS": re.compile(
        r"\b(?:(?:25[0-5]|2[0-4]\d|[01]?\d\d?)\.){3}(?:25[0-5]|2[0-4]\d|[01]?\d\d?)\b"
    ),
    "IBAN_CODE": re.compile(r"\b[A-Z]{2}\d{2}[A-Z0-9]{11,30}\b"),
}

DEFAULT_PII_TYPES = list(_REGEX_PATTERNS.keys())


_config: dict[str, Any] = {
    "enabled": False,
    "pii_types": DEFAULT_PII_TYPES,
    "replacement_format": "[{type}]",
}


def _setup_presidio() -> None:
    global _analyzer, _anonymizer
    if _presidio_available and _analyzer is None:
        _analyzer = AnalyzerEngine()
        _anonymizer = AnonymizerEngine()


def _make_replacement(pii_type: str) -> str:
    return _config["replacement_format"].replace("{type}", pii_type)


def _redact_regex(text: str, pii_types: list[str]) -> str:
    result = text
    for pii_type in pii_types:
        pattern = _REGEX_PATTERNS.get(pii_type)
        if pattern:
            result = pattern.sub(_make_replacement(pii_type), result)
    return result


def _redact_presidio(text: str, pii_types: list[str]) -> str:
    _setup_presidio()
    if _analyzer is None or _anonymizer is None:
        return _redact_regex(text, pii_types)
    results = _analyzer.analyze(text=text, entities=pii_types, language="en")
    if not results:
        return text
    operators = {
        r.entity_type: OperatorConfig(
            "replace", {"new_value": _make_replacement(r.entity_type)}
        )
        for r in results
    }
    return _anonymizer.anonymize(
        text=text, analyzer_results=results, operators=operators
    ).text


def redact(text: str, *, pii_types: list[str] | None = None) -> str:
    """Redact PII from text. Uses Presidio if available, else regex."""
    types = pii_types or _config["pii_types"]
    if _presidio_available:
        return _redact_presidio(text, types)
    return _redact_regex(text, types)


def _redact_dict(obj: Any) -> Any:
    """Recursively redact PII from all string values in a dict/list."""
    if isinstance(obj, str):
        return redact(obj)
    if isinstance(obj, dict):
        return {k: _redact_dict(v) for k, v in obj.items()}
    if isinstance(obj, list):
        return [_redact_dict(item) for item in obj]
    return obj


def configure(
    *,
    enabled: bool = True,
    pii_types: list[str] | None = None,
    replacement_format: str | None = None,
) -> None:
    """Configure auto-mode. Called by jamjet.configure(redact=True)."""
    _config["enabled"] = enabled
    if pii_types is not None:
        _config["pii_types"] = pii_types
    if replacement_format is not None:
        _config["replacement_format"] = replacement_format

File: sdk/typescript/packages/cloud/src/redaction.ts
const REGEX_PATTERNS: Record<string, RegExp> = {
  EMAIL_ADDRESS: /[a-zA-Z0-9._%+\-]+@[a-zA-Z0-9.\-]+\.[a-zA-Z]{2,}/g,
  CREDIT_CARD: /\b(?:\d[\s\-]?){13,15}\d\b/g,
  US_SSN: /\b\d{3}[-\s]?\d{2}[-\s]?\d{4}\b/g,
  PHONE_NUMBER: /(?<!\w)(?:\+?1[-.\s]?)?\(?\d{3}\)?[-.\s]?\d{3}[-.\s]?\d{4}(?!\w)/g,
  IP_ADDRESS: /\b(?:(?:25[0-5]|2[0-4]\d|[01]?\d\d?)\.){3}(?:25[0-5]|2[0-4]\d|[01]?\d\d?)\b/g,
  IBAN_CODE: /\b[A-Z]{2}\d{2}[A-Z0-9]{11,30}\b/g,
}

export const DEFAULT_PII_TYPES = [
  'EMAIL_ADDRESS',
  'CREDIT_CARD',
  'US_SSN',
  'PHONE_NUMBER',
  'IP_ADDRESS',
  'IBAN_CODE',
] as const

export type PiiType = (typeof DEFAULT_PII_TYPES)[number]

export type RedactOptions = {
  piiTypes?: readonly string[]
  replacementFormat?: string
}

function makeReplacement(piiType: string, format: string): string {
  return format.replace('{type}', piiType)
}

export function redact(text: string, opts: RedactOptions = {}): string {
  const types = opts.piiTypes ?? DEFAULT_PII_TYPES
  const format = opts.replacementFormat ?? '[{type}]'
  let result = text
  for (const piiType of types) {
    const pattern = REGEX_PATTERNS[piiType]
    if (pattern) {
      result = result.replace(pattern, makeReplacement(piiType, format))
    }
  }
  return result
}

export function redactDict(obj: unknown, opts: RedactOptions = {}): unknown {
  if (typeof obj === 'string') return redact(obj, opts)
  if (Array.isArray(obj)) return obj.map((item) => redactDict(item, opts))
  if (obj !== null && typeof obj === 'object') {
    const out: Record<string, unknown> = {}
    for (const [k, v] of Object.entries(obj)) {
      out[k] = redactDict(v, opts)
    }
    return out
  }
  return obj
}
